packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -0,0 +1,125 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {Result} from '../../api/search/search/result';
+import {isArray, removeDuplicates} from '../../utils/utils';
+import {executeSearch, fetchMoreResults} from '../search/search-actions';
+import {registerFolding} from './folding-actions';
+import {
+  FoldedResult,
+  FoldingFields,
+  getFoldingInitialState,
+} from './folding-state';
+
+interface ResultWithFolding extends Result {
+  parentResult?: ResultWithFolding;
+  childResults?: ResultWithFolding[];
+}
```

Comment:
I could add those to `Result`, but they are only needed here and I feel like they would be confusing otherwise.

Reply:
> I don't like that these properties are optional

The API sometimes returns both and sometimes a single one of the two.

> I don't like [...] that they are at the same level as the `Result` object

In the case of `ResultWithFolding`, those are returned by the API, therefor we can't choose the level at which they are.

> Rather than extending `Result`, could we build the object like this?

We can't build `ResultWithFolding` like that, but we can build `FoldedResult` like that.

Given your comments, I realize that `FoldedResult` and `ResultWithFolding` are confusing due to their naming and both extending `Result`, so I'll take your suggestion for `FoldedResult`, which should make the distinction much simpler.

Reply:
I see, so if `ResultWithFolding` is describing the api, then having the properties at the same level is correct.
Regarding the api though, why does it sometimes return both and sometimes return one or the other? Is it one or the other at the leaf nodes of the tree perhaps? Or is there something else going on?

But yes, we should manage that in the `FoldedResult` interface exposed to clients.

Reply:
hmm, looking back at the JSON response, it looks like I'm wrong. Only `raw.foldingparent` and `raw.foldingchild` are optional. `parentResult` is nullable but `childResults` is always set (although it can be an empty array).

`parentResult` is null when there's no parent (for the root).

Reply:
update: concerning using an interface that doesn't extend Result, that's unfortunately not as compatible with the idea of making the `Folding` controller into `FoldedResultList`.

By making `FoldedResultList` extend `ResultList`, we may as well change the type of `FoldedResultList.result` to be `FoldedResult`.